src/components/tx/ApprovalEditor/index.tsx

```diff
@@ -0,0 +1,96 @@
+import TokenIcon from '@/components/common/TokenIcon'
+import type { BaseTransaction } from '@safe-global/safe-apps-sdk'
+import { WarningOutlined } from '@mui/icons-material'
+import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
+
+import { Accordion, AccordionDetails, AccordionSummary, IconButton, Skeleton, Typography } from '@mui/material'
+import type { DecodedDataResponse } from '@safe-global/safe-gateway-typescript-sdk'
+import { groupBy } from 'lodash'
+import css from './styles.module.css'
+import { UNLIMITED_APPROVAL_AMOUNT } from '@/utils/tokens'
+import { ApprovalEditorForm } from './ApprovalEditorForm'
+import { useMemo } from 'react'
+import { type ApprovalInfo, APPROVAL_SIGNATURE_HASH, extractTxs, updateApprovalTxs } from './utils/approvals'
+import { useApprovalInfos } from './hooks/useApprovalInfos'
+
+const Summary = ({ approvalInfos, uniqueTokenCount }: { approvalInfos: ApprovalInfo[]; uniqueTokenCount: number }) => {
+  if (approvalInfos.length === 1) {
+    const approval = approvalInfos[0]
+    const amount = UNLIMITED_APPROVAL_AMOUNT.eq(approval.amount) ? 'Unlimited' : approval.amountFormatted
+    return (
+      <Typography display="inline-flex" alignItems="center" gap={1}>
+        <WarningOutlined color="warning" />
+        Give access to <b>{amount}</b>
+        <TokenIcon logoUri={approval.tokenInfo?.logoUri} tokenSymbol={approval.tokenInfo?.symbol} />
+        {approval.tokenInfo?.symbol}
+      </Typography>
+    )
+  }
+
+  return (
+    <Typography color="warning" display="inline-flex" alignItems="center" gap={1}>
+      <WarningOutlined color="warning" />
+      Give access to{' '}
+      <Typography color="warning.main">
+        {uniqueTokenCount} Token{uniqueTokenCount > 1 ? 's' : ''}
+      </Typography>
+    </Typography>
+  )
+}
+
+export const ApprovalEditor = ({
+  txs,
+  updateTxs,
+}: {
+  txs: BaseTransaction[] | (DecodedDataResponse & { to: string })
+  updateTxs?: (txs: BaseTransaction[]) => void
+}) => {
+  const extractedTxs = useMemo(() => extractTxs(txs), [txs])
+  const approvalTxs = useMemo(
+    () => extractedTxs.filter((tx) => tx.data.startsWith(APPROVAL_SIGNATURE_HASH)),
+    [extractedTxs],
+  )
+
+  const [approvalInfos, _, loading] = useApprovalInfos(approvalTxs)
+
+  if (approvalTxs.length === 0) {
+    return null
+  }
+
+  const uniqueTokens = groupBy(approvalTxs, (tx) => tx.to)
+  const uniqueTokenCount = Object.keys(uniqueTokens).length
```

Comment:
This can be moved into the `Summary` component and it need not be passed in. It might be worthwhile memoising the `groupBy` call as well.